front/ui/storybook/stories/ui-core/ComboBoxCustomSuggestions.stories.tsx

```diff
@@ -0,0 +1,454 @@
+  if (!query) {
+    return allSuggestions.map((suggestion) => ({
+      ...suggestion,
+      chList: suggestion.chList.map((ch) => ({
+        ...ch,
+        isCandidate: true,
+        isBestSuggestion: false,
+      })),
+    }));
+  }
+
+  const tokens = query.split(/[\s]+/).filter(String);
+
+  if (endsWithSpace && tokens.length === 1) {
+    const tokenUpper = tokens[0].toUpperCase();
+    const exactTrigramMatches = allSuggestions.filter(
+      (s) => s.trigram.toUpperCase() === tokenUpper
+    );
+
+    if (exactTrigramMatches.length > 0) {
+      return exactTrigramMatches.map((s) => ({
+        ...s,
+        chList: s.chList.map((ch) => ({
+          ...ch,
+          isCandidate: true,
+          isBestSuggestion: false,
+        })),
+      }));
+    }
+  }
+
+  // CASE 1 : only one word typed
+  if (tokens.length === 1) {
+    const token = tokens[0];
+    const tokenNorm = normalize(token);
+    const tokenUpper = token.toUpperCase();
+
+    const scored = allSuggestions
+      .map((suggestion) => {
+        const nameNorm = normalize(suggestion.name);
+        const trigramUpper = suggestion.trigram.toUpperCase();
+
+        const matchesName = nameNorm.includes(tokenNorm);
+        const matchesTrigram = trigramUpper.includes(tokenUpper);
+        const matchesCh = suggestion.chList.some((ch) =>
+          ch.code.toUpperCase().includes(tokenUpper)
+        );
+
+        if (!matchesName && !matchesTrigram && !matchesCh) return null;
+
+        const trigramExact = trigramUpper === tokenUpper;
+        const trigramPrefix = !trigramExact && trigramUpper.startsWith(tokenUpper);
+        const namePrefix = nameNorm.startsWith(tokenNorm);
+        const nameContains = matchesName && !namePrefix;
+        const chPrefix = suggestion.chList.some((ch) =>
+          ch.code.toUpperCase().startsWith(tokenUpper)
+        );
+        const chContains = matchesCh && !chPrefix;
+
+        let category = 6;
+        if (trigramExact) category = 0;
+        else if (trigramPrefix) category = 1;
+        else if (namePrefix) category = 2;
+        else if (nameContains) category = 3;
+        else if (chPrefix) category = 4;
+        else if (chContains) category = 5;
+
+        return { suggestion, category };
+      })
+      .filter((entry) => entry !== null);
+
+    const ordered = scored
+      .sort(
+        (a, b) =>
+          a.category - b.category || a.suggestion.name.localeCompare(b.suggestion.name, 'fr')
+      )
+      .map((entry) => entry.suggestion);
+
+    return ordered.map((suggestion) => ({
+      ...suggestion,
+      chList: suggestion.chList.map((ch) => ({
+        ...ch,
+        isCandidate: true,
+        isBestSuggestion: false,
+      })),
+    }));
+  }
+
+  // CASE 2 : several words
+  const baseText = tokens.slice(0, -1).join(' ');
+  const lastToken = tokens[tokens.length - 1];
+
+  const baseNorm = normalize(baseText);
+  const baseUpper = baseText.toUpperCase();
```

Comment:
You will use this for the trigram, so you can use only the first token
```suggestion
  const baseUpper = tokens[0].toUpperCase();
```